Make s3 client ops use abort source + use in backup task

elcallio opened the pull request:
Fixes #20716

Adds optional abort_source to all s3 client operations. If provided, will propagate to actual HTTP client and allow for aborting actual net op.
    
Note: this uses an abort source per call, not a client-local one.
This is for two reasons:
    
1.) The usage pattern of the client object is to create it outside the eventual owning object (task) that hosts the relevant abort source
2.) It is quite possible to want to have different/no abort source for some operation usage.

Also adds forward usage of task abort_source in backup tasks upload s3 call, making it more readily abort-able.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12963/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/81/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1525/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12625/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4599/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12963/)
#### Build Details:
- Duration: 6 hr 18 min
- Builder: i-0be8916bacc433b16 (m5ad.12xlarge)

tchaikov reviewed (approved):
I'm wondering how we plan to test the finer-grained abort mechanism using/reusing `test_backup_is_abortable()`. Currently, this test injects a breakpoint to pause and abort the backup after a file upload. However, this new changeset aims to abort the upload mid-request by signaling the `abort_source` to shut down the underlying HTTP connection.

I realize this could be a tricky scenario to test. While it's not a blocker for this change, still wanted to hear your thoughts on potential strategies.

xemul commented on utils/s3/client.cc:
```diff
@@ -237,37 +237,43 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
-    auto& gc = find_or_create_client();
-    return gc.http.make_request(std::move(req), std::move(handle), expected).handle_exception([] (auto ex) {
+future<> client::do_make_request(group_client& gc, http::request req, http::experimental::client::reply_handler handle, std::optional<http::reply::status_type> expected_opt, seastar::abort_source* as) {
+    auto expected = expected_opt.value_or(http::reply::status_type::ok);
```
Why is this conversion needed? Why not just pass expected_opt as it down to gc.http.make_request() it accepts optional<status_type> naturally

elcallio replied:
Mainly because http::result::status_type::ok is not clearly specified as the default when nullopt is provided. In fact, if arg to http client is nullopt  the result of the call is ignored.

xemul commented on utils/s3/client.cc:
```diff
@@ -237,37 +237,43 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
-    return gc.http.make_request(std::move(req), std::move(handle), expected).handle_exception([] (auto ex) {
+future<> client::do_make_request(group_client& gc, http::request req, http::experimental::client::reply_handler handle, std::optional<http::reply::status_type> expected_opt, seastar::abort_source* as) {
+    auto expected = expected_opt.value_or(http::reply::status_type::ok);
+    return (as
```
(It's not a review comment, but a side note)
The http::client code does backward conversion :( of abort_source reference to pointer, because it's easier for it to work with pointer internally

regevran commented:
As @tchaikov said - tests are missing.

tchaikov commented:
> As @tchaikov said - tests are missing.

please note, i deliberately avoid saying this. again, i wanted to discuss possible ways to implement a test.

regevran commented:
> > As @tchaikov said - tests are missing.
> 
> please note, i deliberately avoid saying this. again, i wanted to discuss possible ways to implement a test.

Sure. However, this PR cannot continue without tests.

xemul commented:
> ... Maybe it is better to equip (i.e. member variable) the client with an abort source that is used across all operations.

Similar was discussed here: https://github.com/scylladb/seastar/pull/2410#discussion_r1735047485

xemul commented:
> > As @tchaikov said - tests are missing.
> 
> please note, i deliberately avoid saying this. again, i wanted to discuss possible ways to implement a test.

We have `object_store/test_backup::test_backup_is_abortable` to extend

tchaikov commented:
> test_backup_is_abortable

yes. see https://github.com/scylladb/scylladb/pull/21431#pullrequestreview-2414743617 . i thought about adding random abort right after the backup API call. but it could be fragile, because, if the amount of data is relatively small, the upload could have finished when we abort the task. and the breakpoint stalls the task at a granularity of file, which might not be enough.

xemul commented:
It's not necessarily to explicitly check that "aborting in the middle of file uploading" works. We already have test for http client request making being abortable in the middle of any stage (in seastar). Next step is to test  for s3 client multipart upload being abortable (related to #20425 and #19533). The existing test for uploading loop being abortable completes the testing plan. What do you think?

elcallio commented on utils/s3/client.cc:
```diff
@@ -237,37 +237,43 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
+future<> client::make_request(http::request req, http::experimental::client::reply_handler handle, std::optional<http::reply::status_type> expected, seastar::abort_source* as) {
+    authorize(req);
+    auto& gc = find_or_create_client();
+    return do_make_request(gc, std::move(req), std::move(handle), expected);
```
Nope. My mistake.

elcallio replied:
Fixed.

elcallio commented:
I purposefully avoided augmenting the test because as @tchaikov said, it is very flakey trying to make any assumption on timing with abort during an upload. And the types don't lend themselves well to unit tests.

elcallio commented:
> I would like to see examples for this. I am not sure.
> To my understanding if you have an s3 client in hand and you are making operations on it - you expect the same abilities for each operation.
> Maybe it is better to equip (i.e. member variable) the client with an abort source that is used across all operations.

By definition, as we use the clients as moveable objects that change owner, the abort source changes with the ownership. A type that changes ownership should not have potential references to a temporary owners objects. While you could go around the problem by either using a setter for a member (bad bad bad) or adding a semi-move constructor that also sets abort source, neither is very clean.

elcallio commented:
Ok, added preemptive check in s3 make_quest (because HTTP client does not do it), added an extra injection point, and a second test case to verify we abort in the s3 client if required. 
This is not the greatest test, because we are still mainly verifying pre-empt checks. But at least it checks the control flow.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12998/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12998/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
#### Failed Tests (20/150):
- [test_backup_is_abortable_in_s3_client.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.1)
- [test_backup_is_abortable_in_s3_client.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.10)
- [test_backup_is_abortable_in_s3_client.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.2)
- [test_backup_is_abortable_in_s3_client.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.3)
- [test_backup_is_abortable_in_s3_client.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.4)
- [test_backup_is_abortable_in_s3_client.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.5)
- [test_backup_is_abortable_in_s3_client.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.6)
- [test_backup_is_abortable_in_s3_client.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.7)
- [test_backup_is_abortable_in_s3_client.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.8)
- [test_backup_is_abortable_in_s3_client.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12998/testReport/junit/%28root%29/test_backup/test_backup_is_abortable_in_s3_client_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_backup_is_abortable_in_s3_client.debug.9)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: i-0117bb77d60dc8ef8 (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13021/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13021/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store/test_backup
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/125/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4652/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1577/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12679/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13021/)
#### Build Details:
- Duration: 4 hr 0 min
- Builder: spider7.cloudius-systems.com

xemul commented on test/object_store/test_backup.py:
```diff
@@ -107,10 +107,9 @@ async def test_backup_to_non_existent_bucket(manager: ManagerClient, s3_server):
-@skip_mode('release', 'error injections are not supported in release mode')
-async def test_backup_is_abortable(manager: ManagerClient, s3_server):
-    '''check that backing up a snapshot for a keyspace works'''
+async def do_test_backup_abort(manager: ManagerClient, s3_server, 
```
Parameters for pytest are usually added by `@pytest.mark.parametrize` fixture. I'm not proposing to make v3 of this PR, but would be good to have such a followup

elcallio replied:
The params to actual tests are @pytest.parameterize. This is a helper routine.
